lib/vrl/lookup/src/lookup_v2/mod.rs

```diff
@@ -1,3 +1,86 @@
+//! | `.baz` | `["a", "b", "c"]`
+//!
+//!
+//! # Traits
+//! There are 2 main traits that define a path. Most functions that use a path for querying
+//! will require one of these traits, rather than a concrete type.
+//!
+//! - [ValuePath] is a path that points to a field inside of a `Value`.
+//! - [TargetPath] is a path that points to a field inside of a `target`. A `target` in VRL refers to
+//! the external data being processed by a VRL script. A `target` has two main sections that can be
+//! pointed to, `event` and `metadata`.  [TargetPath::prefix] identifies the section, and
+//! [TargetPath::value_path] is a [ValuePath] pointing into that section.
+//!
+//! Note that for performance reasons, since [ValuePath] and [TargetPath] require [Clone], these
+//! traits are only implemented on types that are cheap to clone (usually references). That means
+//! when passing in a value (e.g. [OwnedValuePath]) into a function that requires `impl ValuePath`,
+//! it will generally need to be passed in as a reference.
+//!
+//! # Owned Paths
+//! [OwnedValuePath] and [OwnedTargetPath] are pre-parsed paths. That means that accessing fields
+//! using an owned path is very fast. There is an upfront cost however, since owned paths are parsed
+//! when they are created, and the segments are heap allocated. Owned paths should be preferred
+//! if they can be created when performance isn't as much of a concern (e.g. startup time)
+//! and they can be stored for re-use.
+//! Owned paths tend to be easier to work with since you can directly access / manipulate the
+//! segments that make up the path.
+//!
+//! If a path is being created and will only be used once, it may make sense to use other types.
+//! For example here are two different ways to append a segment to a [OwnedValuePath]  before querying
+//! a `Value`:
+//! - Use [OwnedValuePath::with_field_appended] to create a new [OwnedValuePath] and use that. This
+//! method is preferred if the new path will be used multiple times and the path adjustment can be
+//! done in a non performance-critical part of the code (e.g. at startup).
+//! - Use [ValuePath::concat] which con concatenate two [ValuePath]'s very efficiently without
+//! allocating on the heap.
+//!
+//! To convert a string into an owned path, use either [parse_value_path] or [parse_target_path].
+//!
+//! # String Paths
+//! [ValuePath] and [TargetPath] are implemented for [&str]. That means a raw / unparsed string can
+//! be used as a path. This use is discouraged, and may be removed in the future. It mostly
+//! exists for backwards compatibility in places where String paths are used instead of owned paths.
+//! Using string paths is slightly slower than using an owned path (it's still very fast),
+//! but it's easy to introduce bugs since some compile-time type information is missing, such as whether it's a
+//! target vs value path, or if the entire string is meant to be treated as a single segment vs
+//! being parsed as a path.
```

Comment:
I think this reads better..

```suggestion
//! Using string paths is slightly slower than using an owned path. It's still very fast
//! but it is easy to introduce bugs since some compile-time type information is missing -
//! such as whether it is a target vs value path, or if the entire string is meant
//!  to be treated as a single segment vs being parsed as a path.
```